### ISSUE TITLE
Constants in `macro_rules!` aren't hygienic

### ISSUE THREAD

**Issue by dylni** [rust-lang/rust#75420]
I tried this code:

```rust
macro_rules! const_macro {
    ( $x:expr ) => {{
        const X: () = $x;
        X
    }};
}

const X: () = ();
const _: () = const_macro!(X);
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6a373c7bb61038f44867919a753db411

I expected to see this happen: the code compiles without warnings

Instead, this happened:

```
warning: constant is never used: `X`
 --> src/main.rs:8:1
  |
8 | const X: () = ();
  | ^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

error[E0391]: cycle detected when normalizing `_::X`
  |
note: ...which requires const-evaluating + checking `_::X`...
 --> src/main.rs:3:9
  |
3 |         const X: () = $x;
  |         ^^^^^^^^^^^^^^^^^
...
9 | const _: () = const_macro!(X);
  |               --------------- in this macro invocation
note: ...which requires const-evaluating + checking `_::X`...
 --> src/main.rs:3:9
  |
3 |         const X: () = $x;
  |         ^^^^^^^^^^^^^^^^^
...
9 | const _: () = const_macro!(X);
  |               --------------- in this macro invocation
note: ...which requires const-evaluating `_::X`...
 --> src/main.rs:3:9
  |
3 |         const X: () = $x;
  |         ^^^^^^^^^^^^^^^^^
...
9 | const _: () = const_macro!(X);
  |               --------------- in this macro invocation
  = note: ...which again requires normalizing `_::X`, completing the cycle
note: cycle used when const-evaluating `_`
 --> src/main.rs:9:1
  |
9 | const _: () = const_macro!(X);
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

### Meta

`rustc --version --verbose`:
```
rustc 1.45.2 (d3fb005a3 2020-07-31)
binary: rustc
commit-hash: d3fb005a39e62501b8b0b356166e515ae24e2e54
commit-date: 2020-07-31
host: x86_64-unknown-linux-gnu
release: 1.45.2
LLVM version: 10.0
```

---

This code compiles without warnings, because it uses `let` instead of `const`, which is hygienic:

```rust
macro_rules! let_macro {
    ( $x:expr ) => {{
        let x: () = $x;
        x
    }};
}

fn main() {
    let x: () = ();
    let _: () = let_macro!(x);
}
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a414c7bc73d19c9c7d5d7d2b2880335c

**Comment by petrochenkov** [comment]
This is by design, only local variables, labels and `$crate` are hygienic in `macro_rules`.

Macros 2.0 (`macro`) items are supposed to change the hygiene rules by making everything hygienic, but they are unstable and still far from being stabilized.

**Comment by dylni** [comment]
Thanks @petrochenkov. Is there anywhere I can read about why that decision was made, or is it just because macros 2.0 is being developed?

**Comment by petrochenkov** [comment]
@dylni 
Hygiene for items (including constants) is hard to implement, and it wasn't implemented in the Rust 1.0 time frame when `macro_rules` were stabilized. So `macro_rules` were stabilized as is, with partial hygiene, with an eye on implementing the full hygiene in macros 2.0.

**Comment by dylni** [comment]
Thanks. I didn't consider their similarity to items, so that makes sense
